pkg/core/sync.go

```diff
@@ -0,0 +1,210 @@
+	}
+
+	if res.State == Upgraded || res.State == Downgraded {
+		if err := dropProvider(ctx, storage.DSN(), rp, s.ResourceTables); err != nil {
+			return nil, diag.FromError(fmt.Errorf("drop provider failed: %w", err), diag.INTERNAL)
+		}
+		if err := installProvider(ctx, storage.DSN(), db, want, s.ResourceTables); err != nil {
+			return nil, diag.FromError(fmt.Errorf("install provider failed: %w", err), diag.INTERNAL)
+		}
+	}
+
+	log.Debug().Stringer("provider", opts.Provider).Stringer("state", res.State).Msg("provider sync complete")
+	return res, nil
+}
+
+func Drop(ctx context.Context, storage database.Storage, pm *plugin.Manager, provider registry.Provider) diag.Diagnostics {
+	log.Warn().Stringer("provider", provider).Msg("dropping provider schema")
+	s, diags := GetProviderSchema(ctx, pm, &GetProviderSchemaOptions{Provider: provider})
+	if diags.HasDiags() {
+		return diags
+	}
+	log.Info().Str("provider", provider.Name).Str("version", provider.Version).Msg("dropping provider tables")
+	err := dropProvider(ctx, storage.DSN(), provider, s.ResourceTables)
+	if err != nil {
+		return diag.FromError(err, diag.DATABASE, diag.WithSummary("failed to drop provider"))
+	}
+	return nil
+}
+
+func dropProvider(ctx context.Context, dsn string, provider registry.Provider, resourceTables map[string]*schema.Table) error {
+	dType, _, err := sdkdb.ParseDialectDSN(dsn)
+	if err != nil {
+		return err
+	}
+
+	m, err := migrator.New(logging.NewZHcLog(&log.Logger, "migrator"), dType, nil, dsn, fmt.Sprintf("%s_%s", provider.Source, provider.Name))
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := m.Close(); err != nil {
+			log.Error().Err(err).Msg("failed to close migrator connection")
+		}
+	}()
+
+	if err := m.DropProvider(ctx, resourceTables); err != nil {
+		return err
+	}
+
+	sta, err := state.NewMigratedClient(ctx, dsn, logging.NewZHcLog(&log.Logger, "sync"))
+	if err != nil {
+		return diag.FromError(fmt.Errorf("state failed: %w", err), diag.INTERNAL)
+	}
+	defer sta.Close()
+
+	if err := sta.UninstallProvider(ctx, provider); err != nil {
+		return diag.FromError(fmt.Errorf("state failed: %w", err), diag.INTERNAL)
+	}
+
+	return nil
+}
+
+func installProvider(ctx context.Context, dsn string, db execution.QueryExecer, provider *state.Provider, resourceTables map[string]*schema.Table) error {
+	logger := logging.NewZHcLog(&log.Logger, "sync-install")
+
+	tc := migration.NewTableCreator(logger, schema.PostgresDialect{})
+
+	for _, t := range sort.StringSlice(funk.Keys(resourceTables).([]string)) {
+		up, _, err := tc.CreateTableDefinitions(ctx, resourceTables[t], nil)
+		if err != nil {
+			return diag.FromError(fmt.Errorf("CreateTableDefinitions failed for %s: %w", t, err), diag.INTERNAL)
+		}
+		for _, sql := range up {
+			if err := db.Exec(ctx, sql); err != nil {
+				return diag.FromError(fmt.Errorf("error creating table %s: %w", t, err), diag.INTERNAL)
+			}
+		}
```

Comment:
if any table creation fails, maybe we want to try to create all of them and collect all failed?

Reply:
it will be transactional later on, and won't break the schema.

Reply:
still, even if its transactional, you want to run them all and then rollback, if two tables are problematic, I rather get two errors, then 1 error, fix, another error.

Reply:
Fixed, but in a tx after a failed query, all following queries would also fail:

```
cq=# begin;
BEGIN
cq=*# drop table aws_non_table;
ERROR:  table "aws_non_table" does not exist
cq=!# select count(1) from aws_ec2_instances;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
cq=!# rollback;
ROLLBACK
```